Cpu feature detection

marshallpierce opened the pull request:
As per #357. It looked feasible to re-implement basically all of the offending C file, so I did so.

I've sprinkled TODOs where there are things still to resolve, but the main things I'd like your input on are:

- Linking `getauxval`: In C, it's linked weakly and used if it's available at runtime. I couldn't find how to do this in Rust. Is there a way, or is the best option to leave a weak linkage in C with a little helper function that we call from Rust to then invoke `getauxval` if available and otherwise return an error?
- How should I actually try running this? I don't have an ARM linux system lying around, so the stuff that actually wires the globals, etc in `cpu_feature.rs` is just a guess at how things might look.
- Searching for `GFp_has_broken_NEON` yields no usages. Can we ditch that function and its underlying `g_has_broken_neon` global? (`GFp_armcap_P` is used all over, so the logic is still used to populate that.)
- `getauxval` (and `/proc/self/auxv`) are defined in terms of `unsigned long`. To get things working I used `libc`'s `c_ulong`, but IIRC you're trying to stay `libc`-free. How should I represent `unsigned long` in its many varieties since #380 is still extant?

marshallpierce commented:
Oh, and I've had to liberally apply `#[allow(dead_code)]` because ultimately this is all invoked from an arm-only hook. Is that just how this has to be, or is there something cleverer?

briansmith commented on src/cpu_feature/auxv.rs:
```diff
@@ -0,0 +1,159 @@
+        let found_aux_type = reader.read_u64::<NativeEndian>()
+            .map_err(|_| AuxValError::InvalidFormat)?;
+        let aux_val = reader.read_u64::<NativeEndian>()
+            .map_err(|_| AuxValError::InvalidFormat)?;
```
How terrible would it be to replace these two lines with lines that don't require the `byteorder` crate?

marshallpierce replied:
Depends how much it is problematic to depend on byteorder... :) 

At first glance, https://github.com/BurntSushi/byteorder/blob/master/src/lib.rs is not enormous, so we could strip that down to only support the 32 bit and 64 bit cases, and then do some manual buffering rather than use their `Read` extension. Several hundred lines in total, assuming we maintain the tests. It certainly can be done, but if we keep anything like the same API it will mean keeping most of that library as an internal copy. 

If avoiding this particular dependency is of non-trivial importance, I can trace through the byteorder code and see if there's a more svelte way to extract what we need that would be more svelte.

briansmith commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+#[cfg(all(target_os = "linux", target_arch = "arm"))]
+#[allow(dead_code)] // TODO
+fn search_procfs_auxv(auxv_types: &[c_ulong]) -> Result<AuxVals, AuxValError> {
+    let mut auxv = File::open("/proc/self/auxv")
```
I suggest that you merge all this logic into `auxv::search_auxv`. In particular, that function should take the path to the file to open, and it should open and read the file itself. Then that function would be used by all the tests, and the arm submodule would simply call `auxv::search_auxv("/proc/self/auxv", <auxv_types>)`.

Doing this would move all the platform-specific code from the cross-platform `cpu_features` module to the target-specific submodule.

marshallpierce replied:
Done. Accepting a `Read` was a misguided attempt to aid testability; it turned out that I never wanted to stitch together blocks of 16 bytes by hand...

briansmith commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+#[cfg(all(target_os = "linux", target_arch = "arm"))]
+#[allow(non_snake_case)]
+extern "C" fn GFp_cpuid_setup() {
+    if let Ok(r) = File::open("/proc/cpuinfo").map(|f| BufReader::new(f)) {
```
See the suggestion below. All of this code is cross-platform except the reference to /proc/cpuinfo, so we should move it all to the cross-platform cpuinfo subcrate, possibly having `parse_cpuinfo` open the file and read it itself.

briansmith commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+// TODO is this used in practice? Its C equivalent is the only user of the
+// global, and the function name has no uses
+#[cfg(all(target_os = "linux", target_arch = "arm"))]
+#[allow(non_snake_case)]
```
In BoringSSL, it is probably used by Google Chrome in order to help decide whether to prefer AES-GCM or ChaCha20-Poly1305 in TLS. We might do something similar later, but for now it is OK to remove it.

marshallpierce replied:
Great, removed.

briansmith commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+use self::arm::arm_cpuid_setup;
+
+#[cfg(all(target_os = "linux", target_arch = "arm"))]
+static HAS_BROKEN_NEON: bool = false;
```
I think you're going to remove this (see below), but note that this would need to be `static mut` so that it can be properly set.

briansmith commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+#[cfg(all(target_os = "linux", target_arch = "arm"))]
+use self::auxv::{AuxValError, AuxVals};
+#[cfg(all(target_os = "linux", target_arch = "arm"))]
+use self::cpuinfo::CpuInfo;
```
I think all the above `use`s can be moved to the `arm` submodule.

marshallpierce replied:
Done; `cpu_feature.rs` is now just a place to host `arm_linux`.

briansmith commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+
+mod auxv;
+mod cpuinfo;
+mod arm;
```
It would be better to name this submodule `arm_linux`. We'll support many non-Linux ARM targets that won't have /proc/cpuinfo or /proc/self/aux, and we should reserve the name `arm` for those non-Linux platforms.

marshallpierce replied:
Done

marshallpierce commented on src/cpu_feature/arm.rs:
```diff
@@ -0,0 +1,281 @@
+use std::collections::HashSet;
+use std::string::{String, ToString};
+
+use super::libc::c_ulong;
```
If you're not worried about maintaining the platform/bitwidth mapping, sounds good to me!

marshallpierce commented on src/cpu_feature/cpu_feature.rs:
```diff
@@ -0,0 +1,61 @@
+}
+
+mod auxv;
+mod cpuinfo;
```
Done, with a slightly different `cfg`; I think the parentheses didn't nest maybe the way you intended? The `target_os` bit ended up excluded from the `all`.

marshallpierce commented on src/cpu_feature/auxv.rs:
```diff
@@ -0,0 +1,159 @@
+        }
+
+        let mut reader = &buf[..];
+        // TODO determine length to read
```
It's referring to the fact that I don't know at runtime whether to call read_u64 or read_u32.

According to libc's implementation of `c_ulong`, solaris is always 64 bits, so using pointer width won't work in that case. Same for `c_long`, by the way -- should we be adding this wrinkle to the declarations in `c.rs`? Redox OS is also always using 64 bits for longs, and that at least used to be 32-bit only OS.

Is there a way we could express via a feature what the width of `ulong` is? That way we could have two method declarations, one that invokes `read_u64` and one with `read_u32`, that are toggled with the feature flag.

briansmith replied:
I see. The important thing is that the auxv code is only used in ARM Linux and we know the ARM Linux ABIs. Therefore, I suggest that instead of using/defining a generic "unsigned long" type in `ring::c`, we “simply” use something `#[cfg(all(target_pointer_width = "32", any(test, all(target_os = "linux", any(target_arch = "arm", target_arch = "aarch64")))))]`  and `#[cfg(all(target_pointer_width = "64", any(test, all(target_os = "linux", any(target_arch = "arm", target_arch = "aarch64")))))]` to define a type alias within this module. The idea is that when you run the tests on any 32-bit target, the 32-bit ARM variant gets used, and when you run the tests on a 64-bit target, the 64-bit AAarch64 variants get used, even when those types don't match the target platforms definition of "unsigned long". Does that make sense?

(The main reason we want to support running these tests on non-ARM-Linux platforms is to make them easy to maintain for people that don't have access to such targets.)

marshallpierce replied:
It ended up being slightly more complicated because even on a 32-bit platform we need the tests to use 64 bits because that's what the test auxv files use. Fortunately, it only incurred a minor syntactic burden in a few places.

marshallpierce commented:
Thanks for the feedback; I think I've addressed everything in round 1. I originally had `auxv` and `cpuinfo` all in one, but it started feeling a little beefy as I was scrolling back and forth writing tests for this or that, so I split it up. If we put it all in `arm_linux.rs` it will be a ~600 line file. To my eyes it's a little more manageable mentally (and scroll-wise) in three files, but I don't feel strongly about it, so I don't mind coalescing them if you prefer.

I don't think I know of any other concrete cpuinfo scenarios that need tests, but I'm happy to add more if you know of other problematic cases.

briansmith commented on src/cpu_feature/arm_linux/arm_linux.rs:
```diff
@@ -0,0 +1,307 @@
+
+    // TODO link to getauxval weakly and use that if available. In particular,
+    // this is needed by the assumption in hwcap_from_cpuinfo that future architectures
+    // will have a working getauxval.
```
For `getauxval()`, I recommend just defining a wrapper around it in C code that returns an error code if the function isn't available.

marshallpierce replied:
Done. I'm not particularly confident if I put the C bits in the best possible place, though.

briansmith commented:
> It ended up being slightly more complicated because even on a 32-bit platform we need the tests to use 64 bits because that's what the test auxv files use. Fortunately, it only incurred a minor syntactic burden in a few places.

I'd prefer that we make some 32-bit ARM auxv files and use them on 32-bit platforms, and similarly get some 64-bit Aarch64 auxv files and use them on 64-bit platforms, and not test 64-bit files on 32-bit and vice versa. We can probably do a bit of crowdsourcing of you're having trouble getting such files.

marshallpierce commented:
Do you want me to scrap the `AuxvUnsignedLong` abstraction, then? Too many generics? :) I can virtualize 32-bit x86 linux and grab that auxv, but I don't have any arm hardware. I could try setting up qemu to emulate arm and run linux on that, and while that's better than nothing, it wouldn't really be representative of troublesome (or not) phone hardware. I searched a bit and did not find any sample auxv files, so if you have a way to get those bytes from users with the relevant phones, that would be great.

marshallpierce commented on src/cpu_feature/arm_linux/arm_linux.rs:
```diff
@@ -0,0 +1,514 @@
+        let mut hwcap2 = AuxvUnsignedLongNative::from(0_u32);
+        if let Some(v) = getauxval_provider.getauxval(auxval_types.AT_HWCAP2) {
+            hwcap2 = v;
+        } else if let Some(v) = procfs_auxv.get(&auxval_types.AT_HWCAP2) {
```
The original C did not check procfs auxv for `HWCAP2` if `getauxval` was unavailable. I can't imagine why we wouldn't check that, so I put it in, but maybe you can think of why that would be undesirable?

briansmith commented on crypto/cpu.c:
```diff
@@ -0,0 +1,19 @@
+#include <openssl/cpu.h>
+
+#ifdef __linux__
```
Yes, that seems right to me. However this is also ARM-specific code. I'd prefer it to be in the cpu-arm-linux.c file that you deleted, to keep that clear.

marshallpierce commented:
Thanks for the detailed feedback. I certainly understand that this PR has become quite hefty, so I'll split it up. I'll leave this one open for a bit so I can continue to follow up on existing comments, but I'll close it once that's done.

marshallpierce commented on crypto/cpu.c:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    unsigned long result = getauxval(type);
+    if (errno != 0) {
```
I agree, this is weird, and felt wrong even with my pretty rusty C. Unfortunately, 0 is a valid value, so the original `getauxval` API was simply broken in that case... In the manpage, they have this:

> BUGS
       Before  the  addition  of  the ENOENT error in glibc 2.19, there was no way to unambiguously distinguish the case where type could not be found from the case where the value corresponding to type was zero.

This is the implementation in glibc's `misc/auxval.c`:
```
unsigned long int
__getauxval (unsigned long int type)
{
#ifdef HAVE_AUX_VECTOR
  ElfW(auxv_t) *p;
#endif

  if (type == AT_HWCAP)
    return GLRO(dl_hwcap);
  else if (type == AT_HWCAP2)
    return GLRO(dl_hwcap2);

#ifdef HAVE_AUX_VECTOR
  for (p = GLRO(dl_auxv); p->a_type != AT_NULL; p++)
    if (p->a_type == type)
      return p->a_un.a_val;
#endif

  __set_errno (ENOENT);
  return 0;
}
```

They also have `elf/tst-auxv.c`:
```
static int
do_test (int argc, char *argv[])
{
  errno = 0;
  const char *execfn = (const char *) getauxval (AT_NULL);

  if (errno != ENOENT)
    {
      printf ("errno is %d rather than %d (ENOENT) on failure\n", errno,
          ENOENT);
      return 1;
    }

  if (execfn != NULL)
    {
      printf ("getauxval return value is nonzero on failure\n");
      return 1;
    }

  errno = 0;
  execfn = (const char *) getauxval (AT_EXECFN);

  if (execfn == NULL)
    {
      printf ("No AT_EXECFN found, AT_EXECFN test skipped\n");
      return 0;
    }

  if (errno != 0)
    {
      printf ("errno erroneously set to %d on success\n", errno);
      return 1;
    }

  if (strcmp (argv[0], execfn) != 0)
    {
      printf ("Mismatch: argv[0]: %s vs. AT_EXECFN: %s\n", argv[0], execfn);
      return 1;
    }

  return 0;
}
```

I took this to mean we *should* check errno. https://www.securecoding.cert.org/confluence/pages/viewpage.action?pageId=6619179 seems to agree that while regrettable, this is how it has to be done when every possible return value is potentially valid. Thoughts?

marshallpierce commented on src/cpu_feature/arm_linux/auxv.rs:
```diff
@@ -0,0 +1,313 @@
+
+    #[test]
+    fn test_parse_auxv_virtualbox_linux_32bit() {
+        let path = Path::new("src/cpu_feature/arm_linux/test-data/macos-virtualbox-linux-x86-4850HQ.auxv");
```
I was unable to find 32-bit arm auxv files, so if you've got any, that would be great!

marshallpierce commented on src/cpu_feature/arm_linux/arm_linux.rs:
```diff
@@ -0,0 +1,514 @@
+    return ret;
+}
+
+fn hwcap_from_cpuinfo<T: AuxvUnsignedLong>(cpuinfo: &CpuInfo,
```
Depends on what you expect as a reader from `cpuinfo`, I guess -- I kept that module completely hardware agnostic. At least in its current form, it only exists to put a pretty face on the format of `/proc/cpuinfo` (just as auxv.rs is only a nice way to access the aux vector), and all arm-specific decisions based on that data are elsewhere. For context, my instinct with this sort of thing is to try to make things general purpose so as to avoid leaking concerns across layers unless there are performance concerns or something like that that would preclude such a design. As it is, you could hoist cpuinfo.rs pretty much as is into its own crate and it wouldn't leak any of its arm-feature-detection heritage, and that feels right to me, but I'm not tied to that structure if it bugs you.

marshallpierce commented on src/cpu_feature/arm_linux/cpuinfo.rs:
```diff
@@ -0,0 +1,153 @@
+
+#[cfg(all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux"))]
+pub fn parse_cpuinfo() -> Result<CpuInfo, CpuInfoError> {
+    match File::open("/proc/cpuinfo").map(|f| BufReader::new(f)) {
```
OK. BTW, do you object to `?` rather than the `try!` macro?

marshallpierce commented on src/cpu_feature/arm_linux/cpuinfo.rs:
```diff
@@ -0,0 +1,153 @@
+/// parse the contents of /proc/cpuinfo into a map of field names to values.
+/// Keeps the first encountered value for each name.
+#[cfg(any(test, target_os="linux"))]
+pub fn parse_cpuinfo_reader(input: &mut BufRead)
```
I could change it to take a Path instead of a BufRead and convert the tests that use string literals to all have their own files instead, but I'll still need *something* that's not just "read from exactly /proc/cpuinfo" to be able to test it.

The reason it's public is that it's used in testing the higher-level armcap logic with various cpuinfo files. Is there a better way to expose that to a test submodule of another module?

briansmith commented:
I'm closing this in favor of #408 and whatever follow-up PRs come from that.

gnzlbg commented on src/cpu_feature/arm_linux/auxv.rs:
```diff
@@ -0,0 +1,313 @@
+
+    #[test]
+    fn test_parse_auxv_virtualbox_linux_32bit() {
+        let path = Path::new("src/cpu_feature/arm_linux/test-data/macos-virtualbox-linux-x86-4850HQ.auxv");
```
@briansmith did you manage to get any 32-bit ARM auxv files? We could use them in stdsimd :)
